dotnet nuget sign doc

heng-liu opened the pull request:
## Summary

dotnet 6.0.0-preview.5 SDK has added dotnet nuget sign command.

Happy to iterate to make the doc better based on feedback.

Fixes https://github.com/dotnet/docs/issues/25049

heng-liu committed:
add document for dotnet nuget sign command

heng-liu committed:
add index

Youssef1313 commented on docs/core/tools/dotnet-nuget-sign.md:
```diff
@@ -0,0 +1,162 @@
+
+  Specifies the file path to the certificate to be used in signing the package.
+
+> [!NOTE]
+> This command currently supports only `PKCS12 (PFX)` files that contain the certificate's private key.
```
Indentation to make it appear under the list item.

```suggestion
  > [!NOTE]
  > This command currently supports only `PKCS12 (PFX)` files that contain the certificate's private key.
```

heng-liu replied:
Thanks! Fixed.

Youssef1313 commented on docs/core/tools/dotnet-nuget-sign.md:
```diff
@@ -0,0 +1,162 @@
+
+- **`--certificate-store-name <STORENAME>`**
+
+  Specifies the name of the X.509 certificate store to use to search for the certificate. Defaults to "My", the X.509 certificate store for personal certificates. This option should be used when specifying the certificate via --certificate-subject-name or --certificate-fingerprint options.
```
```suggestion
  Specifies the name of the X.509 certificate store to use to search for the certificate. Defaults to "My", the X.509 certificate store for personal certificates. This option should be used when specifying the certificate via `--certificate-subject-name` or `--certificate-fingerprint` options.
```

Youssef1313 replied:
"My" may get accidentally localized, consider using no-loc.

heng-liu replied:
May I know if there is any example of using no-loc? Thanks! I'm not sure if I need to add a comment here.

Youssef1313 replied:
@heng-liu There are two ways to specify no-loc:

1. Using metadata, which applies globally to the whole article. Example:
https://github.com/dotnet/docs/blob/4ba9a45082b3f1af761ab5bd386c13c4d3712d53/docs/core/compatibility/aspnet-core/6.0/blazor-eventargstype-property-replaced.md#L6

2. Using inline syntax, which applies to one specific occurrence of the word. Example:
https://github.com/dotnet/docs/blob/6c268f75fbd5e9de9a220de13e9ddc7fb2b568ed/includes/core-changes/aspnetcore/3.0/pubternal-apis-removed.md#L21
https://github.com/dotnet/docs/blob/c7d02a73f5736abdd00f9997ae6e64536d4f6068/docs/core/diagnostics/available-counters.md#L18-L20

heng-liu replied:
Thanks @Youssef1313 for the detailed info!
I added no-loc comment. Could you help check if I'm doing right?  Thanks!

Youssef1313 replied:
@heng-liu The syntax looks correct. But I can't confirm on staging since it's available only to MS Employees. Tagging @tdykstra to confirm on staging.

Youssef1313 commented on docs/core/tools/dotnet-nuget-sign.md:
```diff
@@ -0,0 +1,162 @@
+
+- **`--certificate-store-location <STORELOCATION>`**
+
+  Specifies the name of the X.509 certificate store use to search for the certificate. Defaults to "CurrentUser", the X.509 certificate store used by the current user. This option should be used when specifying the certificate via --certificate-subject-name or --certificate-fingerprint options.
```
```suggestion
  Specifies the name of the X.509 certificate store use to search for the certificate. Defaults to "CurrentUser", the X.509 certificate store used by the current user. This option should be used when specifying the certificate via `--certificate-subject-name` or `--certificate-fingerprint` options.
```

Youssef1313 replied:
Same no-loc comment about CurrentUser

heng-liu replied:
Thanks! Could you pls point me to an example of no-loc comment? Thanks!

heng-liu replied:
Thanks! No-loc comment is added :)

Youssef1313 commented on docs/core/tools/dotnet-nuget-sign.md:
```diff
@@ -0,0 +1,162 @@
+
+- **`--certificate-subject-name <SUBJECTNAME>`**
+
+  Specifies the subject name of the certificate used to search a local certificate store for the certificate. The search is a case-insensitive string comparison using the supplied value, which will find all certificates with the subject name containing that string, regardless of other subject values. The certificate store can be specified by --certificate-store-name and --certificate-store-locati ...
```
```suggestion
  Specifies the subject name of the certificate used to search a local certificate store for the certificate. The search is a case-insensitive string comparison using the supplied value, which will find all certificates with the subject name containing that string, regardless of other subject values. The certificate store can be specified by `--certificate-store-name` and `--certificate-store-location` options.
```

heng-liu replied:
Thanks! Fixed.

Youssef1313 commented on docs/core/tools/dotnet-nuget-sign.md:
```diff
@@ -0,0 +1,162 @@
+
+  Specifies the subject name of the certificate used to search a local certificate store for the certificate. The search is a case-insensitive string comparison using the supplied value, which will find all certificates with the subject name containing that string, regardless of other subject values. The certificate store can be specified by --certificate-store-name and --certificate-store-locati ...
+
+> [!NOTE]
+> This command currently supports only a single matching certificate in the result. If there are multiple matching certificates in the result, or no matching certificate in the result, the command will fail.
```
```suggestion
  > [!NOTE]
  > This command currently supports only a single matching certificate in the result. If there are multiple matching 
certificates in the result, or no matching certificate in the result, the command will fail.
```

Youssef1313 replied:
"This command" refers to what? to "--certificate-subject-name"? If so, would "switch" be more suitable than "command" here?

heng-liu replied:
Thanks! Changed into "option".

Youssef1313 commented on docs/core/tools/dotnet-nuget-sign.md:
```diff
@@ -0,0 +1,162 @@
+
+   Specifies the certificate password, if needed. If a certificate is password protected but no password is provided, the command will fail.
+
+> [!NOTE]
+> This command currently supports only noninteractive mode, so there won't be any prompt for a password at run time.
```
```suggestion
  > [!NOTE]
  > This command currently supports only noninteractive mode, so there won't be any prompt for a password at run time.
```

Youssef1313 replied:
Same comment about "command" vs "switch"

heng-liu replied:
Thanks! This is going to say, there is no interactivemode for the sign command. So if the certificate is password protected, but no --certificate-password option is specified, there will be no prompt for a password at run time. (This is different from NuGet.exe command so I think it should be called out)
So changed into: `The sign command currently supports only noninteractive mode, so there won't be any prompt for a password at run time.`

Youssef1313 commented on docs/core/tools/dotnet-nuget-sign.md:
````diff
@@ -0,0 +1,162 @@
+  dotnet nuget sign foo.nupkg --certificate-path cert.pfx --certificate-password password
+  ```
+
+- Sign *foo.nupkg* with certificate(with password protected) matches with the specified SHA-1 fingerprint in the default certificate store(CurrentUser\My):
````
```suggestion
- Sign *foo.nupkg* with certificate (with password protected) matches with the specified SHA-1 fingerprint in the default certificate store (CurrentUser\My):
```

heng-liu replied:
Thanks! Fixed.

Youssef1313 commented on docs/core/tools/dotnet-nuget-sign.md:
````diff
@@ -0,0 +1,162 @@
+  dotnet nuget sign foo.nupkg --certificate-fingerprint 89967D1DD995010B6C66AE24FF8E66885E6E03A8 --certificate-password password
+  ```
+
+- Sign *foo.nupkg* with certificate(with password protected) matches with the specified subject name *Test certificate for testing signing* in the default certificate store(CurrentUser\My):
````
```suggestion
- Sign *foo.nupkg* with certificate (with password protected) matches with the specified subject name *Test certificate for testing signing* in the default certificate store (CurrentUser\My):
```

Youssef1313 replied:
"Test certificate for testing signing" should be in a no-loc.

heng-liu replied:
Thanks! Could you pls point me to an example of no-loc comment?

heng-liu replied:
Thanks! No-loc comment is added for "Test certificate for testing signing".

Youssef1313 commented on docs/core/tools/dotnet-nuget-sign.md:
````diff
@@ -0,0 +1,162 @@
+  dotnet nuget sign foo.nupkg --certificate-subject-name "Test certificate for testing signing" --certificate-password password
+  ```
+
+- Sign *foo.nupkg* with certificate(with password protected) matches with the specified SHA-1 fingerprint in the certificate store CurrentUser\Root:
````
```suggestion
- Sign *foo.nupkg* with certificate (with password protected) matches with the specified SHA-1 fingerprint in the certificate store CurrentUser\Root:
```

heng-liu replied:
Thanks! Fixed.

Youssef1313 commented on docs/core/tools/dotnet-nuget-sign.md:
````diff
@@ -0,0 +1,162 @@
+  dotnet nuget sign foo.nupkg c:\mydir\*.nupkg --certificate-path cert.pfx
+  ```
+
+- Sign *foo.nupkg* with certificate *cert.pfx*(with password protected), and timestamp with *http://timestamp.test*:
````
```suggestion
- Sign *foo.nupkg* with certificate *cert.pfx* (with password protected), and timestamp with *http://timestamp.test*:
```

heng-liu replied:
Thanks! Fixed.

tdykstra reviewed (commented):
This looks great, I have a few suggestions for you to consider.

heng-liu committed:
address feedback

heng-liu commented on docs/core/tools/dotnet-nuget-sign.md:
```diff
@@ -0,0 +1,162 @@
+---
+title: dotnet nuget sign command
+description: The dotnet nuget sign command signs all the packages matching the first argument with a certificate. The certificate with the private key can be obtained from a file or from a certificate installed in a certificate store by providing a subject name or a thumbprint.
```
Thanks! Fixed.

heng-liu commented on docs/core/tools/dotnet-nuget-sign.md:
```diff
@@ -0,0 +1,162 @@
+---
+# dotnet nuget sign
+
+**This article applies to:** ✔️ .NET 6.0.0-preview.5 SDK and later versions
```
Thanks! Fixed.

heng-liu commented on docs/core/tools/dotnet-nuget-sign.md:
```diff
@@ -0,0 +1,162 @@
+    [--timestamper <TIMESTAMPINGSERVER>]
+    [--timestamp-hash-algorithm <HASHALGORITHM>]
+    [--overwrite]
+    [-v|--verbosity <LEVEL>]
```
Thanks! Fixed.

heng-liu commented on docs/core/tools/dotnet-nuget-sign.md:
```diff
@@ -0,0 +1,162 @@
+
+- **`package-path(s)`**
+
+  Specifies the file path to the package(s) to be signed. Multiple position arguments can be passed in to verify multiple packages.
```
Thanks! Fixed.

Youssef1313 commented on docs/core/tools/dotnet-nuget-sign.md:
````diff
@@ -0,0 +1,162 @@
+  dotnet nuget sign foo.nupkg c:\mydir\*.nupkg --certificate-path cert.pfx
+  ```
+
+- Sign *foo.nupkg* with certificate *cert.pfx* (password protected), and timestamp with *http://timestamp.test*:
````
```suggestion
- Sign *foo.nupkg* with certificate *cert.pfx* (password protected), and timestamp with `http://timestamp.test`:
```

heng-liu commented on docs/core/tools/dotnet-nuget-sign.md:
```diff
@@ -0,0 +1,162 @@
+
+- **`--overwrite`**
+
+   Switch to indicate if the current signature should be overwritten. By default the command will fail if the package already has a signature.
```
Thanks! Fixed.

heng-liu commented on docs/core/tools/dotnet-nuget-sign.md:
```diff
@@ -0,0 +1,162 @@
+
+## Examples
+
+- Sign *foo.nupkg* with certificate *cert.pfx* (with no password protected):
```
Thanks! Fixed.

heng-liu commented on docs/core/tools/dotnet-nuget-sign.md:
````diff
@@ -0,0 +1,162 @@
+  dotnet nuget sign foo.nupkg --certificate-path cert.pfx
+  ```
+
+- Sign *foo.nupkg* with certificate *cert.pfx* (with password protected):
````
Thanks! Fixed.

heng-liu commented on docs/core/tools/dotnet-nuget-sign.md:
````diff
@@ -0,0 +1,162 @@
+  dotnet nuget sign foo.nupkg --certificate-fingerprint 89967D1DD995010B6C66AE24FF8E66885E6E03A8 --certificate-password password --certificate-store-location CurrentUser --certificate-store-name Root
+  ```
+
+- Sign multiple NuGet packages - *foo.nupkg* and *all .nupkg files in the directory specified* with certificate *cert.pfx* (with no password protected):
````
Thanks! Fixed.

heng-liu commented on docs/core/tools/dotnet-nuget-sign.md:
````diff
@@ -0,0 +1,162 @@
+  dotnet nuget sign foo.nupkg --certificate-path cert.pfx --certificate-password password --timestamper http://timestamp.test
+  ```
+
+- Sign *foo.nupkg* with certificate *cert.pfx* (with no password protected) and save the signed package under specified directory:
````
Thanks! Fixed.

heng-liu commented on docs/core/tools/dotnet-nuget-sign.md:
````diff
@@ -0,0 +1,162 @@
+  dotnet nuget sign foo.nupkg --certificate-path cert.pfx --output c:\signed\
+  ```
+
+- Sign *foo.nupkg* with certificate *cert.pfx* (with no password protected) and overwrite the current signature if the package is already signed:
````
Thanks! Fixed.

heng-liu commented:
Thanks @Youssef1313 and @tdykstra for your feedbacks! 
I've addressed most of the comments, except the non-loc comments. 
May I know if you could point me to an example of non-loc comments? I tried to search in the repo but failed.

Youssef1313 commented:
> May I know if you could point me to an example of non-loc comment? I tried to search in the repo but failed.

https://github.com/dotnet/docs/pull/25050#discussion_r667107788

tdykstra commented:
For no-loc syntax, see also https://docs.microsoft.com/en-us/contribute/markdown-reference#non-localized-strings

heng-liu committed:
add no-loc comments

heng-liu commented:
Thanks @Youssef1313 and @tdykstra for your help and quick response!
All comments are addressed :)

adegeo commented on docs/core/tools/dotnet-nuget-sign.md:
```diff
@@ -0,0 +1,162 @@
+
+   Specifies the certificate password, if needed. If a certificate is password protected but no password is provided, the sign command will fail.
+
+  > [!NOTE]
+  > The sign command currently supports only noninteractive mode, so there won't be any prompt for a password at run time.
```
`non-interactive` is the more common word related to shell/terminal usage.

```suggestion
  > [!NOTE]
  > The `sign` command only supports non-interactive mode. There won't be any prompt for a password.
```

heng-liu replied:
Thanks for reviewing!
Fixed :)

tdykstra commented:
Staging looks good, this PR is ready to merge as soon as the comment from @adegeo is addressed.

tdykstra commented:
Just noticed -- the Options section wasn't put in alphabetic order to match the synopsis.

heng-liu committed:
fix non-interactive

heng-liu commented:
> Just noticed -- the Options section wasn't put in alphabetic order to match the synopsis.

Thanks! Fixed the order in Options section to match the synopsis.

heng-liu committed:
fix

tdykstra reviewed (approved):
Thanks @heng-liu for creating this page and addressing all of the feedback!
